utf8/reader.go

```diff
@@ -0,0 +1,142 @@
+package utf8
+
+import (
+	"fmt"
+	"io"
+	"unicode/utf8"
+
+	"github.com/madlambda/spells/io/bytes"
+)
+
+type (
+	RuneReader struct {
+		r io.RuneReader
+	}
+
+	// ReaderReader implements a runes.Reader from a bytes.Reader interface.
+	ReaderReader struct {
+		r bytes.Reader
+	}
+
+	Error string
+)
+
+const ErrDecode Error = "invalid rune"
+
+// NewReaderReader creates a runes.Reader implementation from an bytes.Reader
+// interface.
+func NewReaderReader(r bytes.Reader) *ReaderReader {
+	return &ReaderReader{
+		r: r,
+	}
+}
+
+// Read implements runes.Reader interface.
+//
+// It will use O(N) space where N is the size of the data slice provided.
+// For ASCII data it allocates exactly len(data) bytes (or sizeof(data)/4) but
+// in case of multi-byte code points it allocates a maximum of sizeof(data)
+// or len(data)*4.
+//
+// The first call to the source reader uses a full buffer of len(data) bytes
+// but subsequent calls read byte by byte. This is a tradeoff optimization for
+// the kind of use cases this function was designed for: parsing languages.
+// Configuration files and programming languages source code are mainly composed
+// of ASCII data, so allocating len(data)*4 seems to be wasteful.
+func (rr *ReaderReader) Read(data []rune) (int, error) {
+	var (
+		nrunes    int
+		runeStart int // decoding starts at this offset
+	)
+
+	end := len(data)
+	b := make([]byte, 0, end)
```

Comment:
```suggestion
	utf8stream := make([]byte, 0, end)
```

But feel free to ignore, I'm extra stupid so I tend to longer names.